### PR TITLE
fix/unittest invalidate on immediate renderer  

### LINKED ISSUE #1998: Controls with Render Transform don't update properly their visuals on ImmediateRenderer
repro steps:

Try use RenderTransform on Button and check it's visual state when mouse over/pressed.

```xaml
     <Button Content="Scaled" HorizontalAlignment="Center" VerticalAlignment="Center">
                        <Button.RenderTransform>
                            <ScaleTransform ScaleX="1.5" ScaleY="1.5" />
                        </Button.RenderTransform>
       </Button>
```

only partial parts of the button is updated when mouse over/pressed etc.

### PR DESCRIPTION
**Pull request by donandren**
issue #1998

This template is not intended to be prescriptive, but to help us review pull requests it would be useful if you included as much of the following information as possible:

- What does the pull request do?
fix/unittest invalidate on immediate renderer   for controls with renderedtransfom 
- What is the current behavior?
ImmediateRenderer is broken, it doesn't invalidate proper area for controls with rendertransform
- What is the updated/expected behavior with this PR?
fix ImmediateRenderer invalidation bounds calculations
- How was the solution implemented (if it's not obvious)?

Checklist:

- [x] Added unit tests (if possible)?

If the pull request fixes issue(s) list them like this:

Fixes #1998

### CODE REVIEW

**Comment by donandren** [comment]
The problem with immediate renderer invalidate is not solved completelly, when animations are performed.
Check ControlCatalog, progressbar page with immediate renderer there are drawing artifacts on the animated progress bars

**Comment by donandren** [comment]
fixes completely #1998  for aminations and controls/containers with rendertransform

File: src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs [hunk 1]
```diff
@@ -115,12 +115,45 @@ public static void Render(IVisual visual, DrawingContext context)
...
+            return result;
+        }
+
         /// <inheritdoc/>
         public void AddDirty(IVisual visual)
         {
             if (visual.Bounds != Rect.Empty)
             {
-                var m = visual.TransformToVisual(_root);
+                var m = TransformToVisual(visual, _root);
```
[the remark this answers is not in the document]
Comment (grokys) [reply src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs]: Yes, I think we need to make `Visual.TransformToVisual` take `RenderTransform` into account. It is [taken into account in WPF](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.uielement.transformtovisual):

> TransformToVisual gives coordinate results after all considerations that affect rendering and positioning such as RenderTransform have been applied. This is useful if you're processing point values that were obtained during an animation of RenderTransform or other position changes.

File: src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs [hunk 2]
```diff
@@ -125,6 +128,13 @@ public void AddDirty(IVisual visual)
                 if (m.HasValue)
                 {
                     var bounds = new Rect(visual.Bounds.Size).TransformToAABB(m.Value);
+
+                    if (_lastPaintBounds != default)
```
Comment (donandren) [inline src/Avalonia.Visuals/Rendering/ImmediateRenderer.cs]: this is not optimal as last Paint is always the full width/height of the window, so essentially we have full frame invalidation every frame.
Essentially we need to invalidate previous bounds (controls) for the bounds (control) that are changed from previous render. So moving progressbar indicator for example invalidates every frame old/moved from and new/moved to bounds

**Review by grokys — APPROVED** [review]

**Pull request merged**